packages/entity-server/src/api/EntityApi.ts

```diff
@@ -0,0 +1,428 @@
+/*
+ * Tupaia
+ * Copyright (c) 2017 - 2021 Beyond Essential Systems Pty Ltd
+ *
+ */
+
+import { OutboundConnection, AuthHandler, PartialOrArray } from '@tupaia/server-boilerplate';
+
+import {
+  CLAUSE_DELIMITER,
+  FIELD_VALUE_DELIMITER,
+  NESTED_FIELD_DELIMITER,
+  MULTIPLE_VALUES_DELIMITER,
+} from '../constants';
+
+import { EntityFields } from '../models';
+import {
+  GroupByAncestorRelationshipsResponseBody,
+  GroupByDescendantRelationshipsResponseBody,
+} from '../routes/hierarchy/relationships';
+import {
+  FlattableEntityFieldName,
+  ExtendedEntityFieldName,
+  EntityResponseObject,
+  FlattenedEntity,
+} from '../routes/hierarchy/types';
+
+type EntityFilterObject = PartialOrArray<EntityFields>;
+
+type EntityApiQueryOptions = {
+  field?: FlattableEntityFieldName;
+  fields?: ExtendedEntityFieldName[];
+  filter?: EntityFilterObject;
+};
+type RelationshipsSubQueryOptions = Omit<EntityApiQueryOptions, 'fields'>;
+
+type EntityApiResponse<T extends ExtendedEntityFieldName[]> = Required<
+  Pick<EntityResponseObject, T[number]>
+>;
```

Comment:
This is another one of my terrifying nested generic types here, but it's a really neat one. The `Pick<EntityResponseObject, T[number]>` part basically says that, given an array of entity field names, the response will be an object that contains each of those fields, and their type will be correct as well. eg.
```
EntityApiResponse<['code', 'childCodes']> = { code: string, childCodes: string[] }
```